src/java.base/share/classes/jdk/internal/foreign/BufferStack.java

```diff
@@ -0,0 +1,188 @@
+     *          allocations by recycling the BufferStacks internal memory}
+     *
+     * @param layout for which to reserve internal memory
+     */
+    @ForceInline
+    public Arena pushFrame(MemoryLayout layout) {
+        return pushFrame(layout.byteSize(), layout.byteAlignment());
+    }
+
+    @Override
+    public String toString() {
+        return "BufferStack[" + size + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(this);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this == obj;
+    }
+
+    private record PerThread(ReentrantLock lock, SlicingAllocator stack, Arena arena) {
+
+        @ForceInline
+        public Arena pushFrame(long size, long byteAlignment) {
+            boolean needsLock = Thread.currentThread().isVirtual() && !lock.isHeldByCurrentThread();
+            if (needsLock && !lock.tryLock()) {
+                // Rare: another virtual thread on the same carrier competed for acquisition.
+                return Arena.ofConfined();
+            }
+            if (!stack.canAllocate(size, byteAlignment)) {
+                if (needsLock) lock.unlock();
+                return Arena.ofConfined();
+            }
+            return new Frame(needsLock, size, byteAlignment);
+        }
+
+        static PerThread of(long size) {
+            final Arena arena = Arena.ofAuto();
+            final SlicingAllocator stack = new SlicingAllocator(arena.allocate(size));
+            return new PerThread(new ReentrantLock(), stack, arena);
+        }
+
+        private final class Frame implements Arena {
+            private final boolean locked;
+            private final long parentOffset;
+            private final long topOfStack;
+            private final Arena confinedArena = Arena.ofConfined();
+            private final SegmentAllocator frame;
+
+            @SuppressWarnings("restricted")
+            @ForceInline
+            public Frame(boolean locked, long byteSize, long byteAlignment) {
+                this.locked = locked;
+                parentOffset = stack.currentOffset();
+                MemorySegment frameSegment = stack.allocate(byteSize, byteAlignment);
+                topOfStack = stack.currentOffset();
+                frame = new SlicingAllocator(frameSegment.reinterpret(confinedArena, null));
+            }
+
+            @ForceInline
+            private void assertOrder() {
+                if (topOfStack != stack.currentOffset())
+                    throw new IllegalStateException("Out of order access: frame not top-of-stack");
+            }
+
+            @ForceInline
+            @Override
+            @SuppressWarnings("restricted")
+            public MemorySegment allocate(long byteSize, long byteAlignment) {
+                // Make sure we are on the right thread
+                if (((MemorySessionImpl) scope()).ownerThread() != Thread.currentThread()) {
```

Comment:
This can also be rewritten as:
`MemorySession.toMemorySession(confinedArena).checkValidState()` -- which is what we use elsewhere